Display correct message error when a stream is already created.

spena opened the pull request:
This fixes https://github.com/confluentinc/ksql/issues/2374

### Description 
If a Stream already exists with the same name, the error message returned uses the term Topic, which is incorrect.

```
ksql> CREATE STREAM PCAP (layers struct<frame struct<frame_frame_interface_id varchar>>) WITH (KAFKA_TOPIC='localPCAP', VALUE_FORMAT='JSON');
Topic already registered: PCAP
```

This patch changes the message to be:
```
ksql> CREATE STREAM PCAP (layers struct<frame struct<frame_frame_interface_id varchar>>) WITH (KAFKA_TOPIC='localPCAP', VALUE_FORMAT='JSON');
Stream already exists: PCAP
```

### Testing done 
One test case added to the `CreateStreamCommandTest` class.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

big-andy-coates commented:
Hey @spena 

So from the github issue it looks like @rmoff would prefer an error message like:

```
Failed to create stream because there is already a stream or table named PCAP registered.
```

So I'm think a good fix might be to change `RegisteredTopicCommand` to throw a better error message. e.g.

```java
@Override
  public DdlCommandResult run(final MetaStore metaStore) {
    final boolean alreadyRegistered = metaStore.getTopic(topicName) != null;
    if (alreadyRegistered) {
      // IF NOT EXIST is set, so this is allowed:
      if (notExists) {
        return new DdlCommandResult(true, "Topic already registered.");
      }

      final String sourceType = getSourceType(metaStore);
      throw new KsqlException(
          "Failed to create because there is already " + sourceType + " named " + topicName
              + " registered.");
    }

    metaStore.putTopic(new KsqlTopic(topicName, kafkaTopicName, topicSerDe, false));

    return new DdlCommandResult(true, "Topic registered");
  }

  private String getSourceType(final MetaStore metaStore) {
    final StructuredDataSource source = metaStore.getSource(topicName);
    if (source == null) {
      return "an entity";
    }

    switch (source.getDataSourceType()) {
      case KSTREAM:
        return "a stream";

      case KTABLE:
        return "a table";

      default:
        return "an entity";
    }
  }
```

spena commented:
@big-andy-coates Thanks for the review. I added the suggested code to the RegisterTopicCommand. I just changed the type a little bit, to show TOPIC if the source is a topic or ENTITY if it is unknown.

spena commented:
This PR is not merged. I don't know what happened (or what I did) to close it. Nothing has changed on the master branch.

rmoff commented on ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/RegisterTopicCommand.java:
```diff
@@ -101,11 +102,10 @@ public DdlCommandResult run(final MetaStore metaStore) {
+        return new DdlCommandResult(true, "Topic already registered.");
       } else {
-        throw new KsqlException("Topic already registered: " + topicName);
+        throw new KsqlException(String.format("Failed to create because %s with name '%s' "
```
Nit: I'd prefer a noun after `create`, at the moment it feels a bit fragmentary. It's clearer than before but whilst we've got the hood up let's nail it :) 

Something like: 

* `Failed to create stream because stream with name x already exists`
* `Failed to create table because stream with name x already exists`

etc

spena replied:
I'd like to do that too, but I would need to add  new parameter to the `RegisterTopicCommand.run() `just to specify the topic type it is trying to create in order to use it in the error message. I don't like that approach just for a custom message.

However, I think I can play with a message thrown by the `RegisterTopicCommand.run()` call and add a new one from the `CreateStreamCommand` and `CreateTableCommand` and make it look better. 

What do you think about this message:

- `Cannot create table 't1': Stream with name 't1' already exists`
- `Cannot create stream 't1': Stream with name 't1' already exists`

If that works, then I can do it in the code like this:

```
run(final MetaStore metaStore) {
  throw KsqlException("%s with name '%s' already exists", sourceType, topicName);
}

...

try {
  run(metastore);
} catch (KsqlException e) {
  throw new KsqlException("Cannot create table '%s': " + e.getMessage(), tableName);
}
```

rmoff replied:
Yes that works for me. I'd want consistent capitalisation of `stream`/`Stream` and `table`/`Table` though ;)
Can you share a sample of the actual output messages once you've tested this?

big-andy-coates replied:
> I'd like to do that too, but I would need to add new parameter to the `RegisterTopicCommand.run() `just to specify the topic type it is trying to create in order to use it in the error message. I don't like that approach just for a custom message.

FYI, a topic can be registered outside of a stream or table being created. So I'm not sure that would work anyway.

spena replied:
I found other issues, though. If I register a topic from ksql using a topic name that I registered from the Confluent UI, then the command works without complaining the topic already existed. Same for streams, if I create a stream from an already registered topic, then it works. 

I will need to fix more places to avoid those situations instead of just fixing the error message. Also, there is a different message when creating a table with a name from a stream:
`Cannot add the new data source. Another data source with the same name already exists: KsqlStream name:STREAM1`

big-andy-coates replied:
Not sure I follow.

By Confluent UI I take it you mean the C3 web UI?   By `Ksql` do you mean the KSQL CLI?

As I've said else where the whole `REGISTER TOPIC` stuff isn't really supported / documented and has bugs and we should rip it out for this very reason.

I'd focus this PR on just fixing issues with CS, CSAS, CT and CTAS statements. 

Feel free to raise follow up PRs to fix other issues or, my preference, rip out `REGISTER TOPIC` functionality,

spena replied:
Yes, by Confluent UI is by using the C3 web UI. I chat with @hjafarpour yesterday btw, and he said there might be future plans to keep the `register topic`. I will leave the generic error message there.

big-andy-coates commented on ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CreateStreamCommandTest.java:
```diff
@@ -134,6 +140,21 @@ public void shouldThrowOnOldWindowProperty() {
+    givenProperties(propsWith(ImmutableMap.of()));
+
+    // When:
+    final CreateStreamCommand cmd = createCmd();
```
It's worth explaining the Given/When/Then thing in case you're not familiar.

Positive tests:

- Given: is the setup for the test, i.e. setting up the pre-conditions.
- When: is the actual _line_ on test.
- Then: is asserting the post conditions.

Negative tests:

- Given: is the setup for the test, i.e. setting up the pre-conditions.
- Expect/Then: the execpted exceptions
- When: is the actual _line_ on test.

So in this test might look like:

```java
// Given:
givenProperties(..)
cmd = createCmd()
cmd.run();

// Expect:
expectedException.blah

// When:
cmd.run();

spena replied:
Thanks, I'll use that way

hjafarpour commented:
@spena instead of creating the message in `RegisterTopicCommand` class you can catch the exception from`RegisterTopicCommand.run()` in `CreateStreamCommand` and `CreateTableCommand` and return the appropriate message there.

big-andy-coates commented on ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CreateTableCommandTest.java:
```diff
@@ -137,6 +142,21 @@ public void shouldThrowOnOldWindowProperty() {
   }
 
+  @Test
+  public void testCreateAlreadyRegisteredTableThrowsException() {
```
Likewise:

```java
    // Given:
    givenProperties(propsWith(ImmutableMap.of()));
    final CreateTableCommand cmd = createCmd();
    cmd.run(metaStore);

    // Then:
    expectedException.expectMessage(
        "Failed to create because a TABLE with name 'name' already exists.");

     // When:
    cmd.run(metaStore);
```

big-andy-coates commented:
> @hjafarpour I left the message on the RegisterTopicCommand, but catch it again in the Stream and Table commands and add a custom message to them. Is that ok?

Works for me.

big-andy-coates commented:
> @rmoff Here are the output examples of how it looks like:

```
ksql> create stream s1 (id int) with (value_format='json', kafka_topic='t1');
Cannot create stream 'S1': An entity with name 'S1' already exists

ksql> create table tab1 (id int) with (value_format='json', kafka_topic='t1', key='ID');
Cannot create table 'TAB1': A table with name 'TAB1' already exists

ksql> create stream tab1 (id int) with (value_format='json', kafka_topic='t1');
Cannot create stream 'TAB1': A table with name 'TAB1' already exists

ksql> create table s1 (id int) with (value_format='json', kafka_topic='t1', key='ID');
Cannot create table 'S1': A stream with name 'S1' already exists

ksql> create stream s2 (id int) with (value_format='json', kafka_topic='t2');
Cannot create stream 'S2': An entity with name 'S2' already exists

ksql> create table s2 (id int) with (value_format='json', kafka_topic='s2', key='ID');
Cannot create table 'S2': An entity with name 'S2' already exists
```

To help me understand, can you add the setup statements to the above as well please? i.e. what statement did you run to get the first statement to return `An entity with name 'S1' already exists`?

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateStreamCommand.java:
```diff
@@ -37,7 +38,12 @@ public CreateStreamCommand(
+      try {
+        registerTopicCommand.run(metaStore);
+      } catch (KsqlException e) {
+        throw new KsqlException(String.format("Cannot create stream '%s': %s",
```
Can you add `e` to the new exception please, e.g. `throw new KsqlException(newMessage, e);`

spena replied:
Done

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateTableCommand.java:
```diff
@@ -60,7 +60,12 @@ public CreateTableCommand(
+      try {
+        registerTopicCommand.run(metaStore);
+      } catch (KsqlException e) {
+        throw new KsqlException(String.format("Cannot create table '%s': %s",
```
As above, add `e` to the thrown exception.

spena replied:
Done

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/RegisterTopicCommand.java:
```diff
@@ -117,4 +118,25 @@ public DdlCommandResult run(final MetaStore metaStore) {
+  private String getSourceType(final MetaStore metaStore) {
+    final StructuredDataSource source = metaStore.getSource(topicName);
+    if (source == null) {
+      return "An entity";
```
This can be `A topic` right?  As `metaStore.getTopic(topicName)` above has returned non-null to get to this point in the code, so we know its a topic.  (You'll notice your example output currently never says `A topic` because that branch of your code may not be reachable).

big-andy-coates replied:
Also, could you add tests to `RegisterTopicCommandTest.java` to cover these different paths?

I know you've added tests to the CreateStream and CreateTable test files, but this doesn't cover all code paths, and really each path should be covered by _this_ classes test file.

big-andy-coates commented on ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CreateStreamCommandTest.java:
```diff
@@ -134,6 +140,23 @@ public void shouldThrowOnOldWindowProperty() {
+
+    // Given:
+    givenProperties(propsWith(ImmutableMap.of()));
+    cmd = createCmd();
```
```suggestion
    final CreateStreamCommand cmd = createCmd();
```

spena replied:
What's wrong with the declaration at the beginning of the method? I think it looks cleaner, but is there a benefit on moving the declaration to this line?

big-andy-coates commented on ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CreateTableCommandTest.java:
```diff
@@ -137,6 +142,23 @@ public void shouldThrowOnOldWindowProperty() {
+
+    // Given:
+    givenProperties(propsWith(ImmutableMap.of()));
+    cmd = createCmd();
```
```suggestion
    final CreateStreamCommand cmd = createCmd();
```

spena commented:
@big-andy-coates The following setup commands were used:

The `An entity with name` error happens when a topic registerd by `register topic` is used. Note that there is no error if the topic exists but not registered by `register topic`.

For instance:

```
ksql> create stream t1 (id int) with (value_format='json', kafka_topic='t1');
works!
ksql> drop stream t1; // clean up to start again

ksql> register topic t1 with (value_format='json',kafka_topic='t1'); 
ksql> create stream t1 (id int) with (value_format='json', kafka_topic='t1');
Cannot create stream 'T1': An entity with name 'T1' already exists
fails! because topic 't1' was registered by ksql previously

ksql> drop topic t1; // clean up to start again
ksql> create stream t1 (id int) with (value_format='json', kafka_topic='t1');
works!
```

The above was very confusing at first, but after chatting with @hjafarpour, I decided not to care about the `register topic`, and just use the create stream and table. However, I made sure that if you register a topic, if I want to register again, then it fails correctly.
